### PR TITLE
Sets up template examples using GMM segmentation class

### PR DESCRIPTION
**Pull request by dtorrejo**
This PR focuses on:
- Cleans up main_methods functions.
- Fixes bug in gmm_segmenter while removing outliers.
- Adds examples on how to use gmm_segmenter.
- Adds watershed segmentation class.
- Adds plots of grains distributions

### CODE REVIEW

File: methods/main_methods.py [hunk 1]
```diff
@@ -13,29 +13,29 @@
 HEIGHT_INDEX = config.height_index
 
 ## Boundary Detection
-def sobel(x):
```
Comment (dtorrejo) [inline methods/main_methods.py:16]: Most of the changes in main_methods change variable names: x -> data, X -> data, O -> outliers, and L -> labels

File: methods/main_methods.py [hunk 2]
```diff
@@ -233,106 +237,105 @@ def show_correlations(num_props, path):
     pyplot.show()
 
 ## Mixture of Gaussians Model
-def segment(X, outliers, num_components=3, normal=True):
+def segment(data, outliers, num_components=2, normal=True):
```
Comment (dtorrejo) [inline methods/main_methods.py]: Simplifies method using Patrick's code as reference

File: methods/main_script.py [hunk 3]
```diff
@@ -4,31 +4,37 @@
...
 # Shows a-priori (classification) property distributions
-mm.show_property_distributions(X)
+mm.show_property_distributions(data, outliers)
 
-# Shows pixel classification after applying Gaussian mixture model
-L, reduced_X = mm.apply_segmentation(X, height_flag=False)
-mm.show_classification(L, reduced_X)
+# Shows pixel classification after applying Gaussian mixture model.
+# NOTE The height_flag keeps/removes the height property when classifying.
+L, reduced_data = mm.apply_segmentation(data, outliers, height_flag=False)
```
Comment (dtorrejo) [inline methods/main_script.py]: @ponl We should add the height_flag to your gmm_segmenter. I don't think the height property field should be using when classifying since it may learn things we want to avoid such as irregularities during experiments.

File: ssts/segmentation/example.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import numpy as np
+import segmentation_gmm as seg_gmm
+from methods import main_methods as mm # NOTE SSTS directory must be in PYTHONPATh
+
+which_example = 2
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: Chooses which example to use

File: ssts/segmentation/example.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+## NOTE Segmentation example without using outliers
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: Simplest case, using default parameters

File: ssts/segmentation/example.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    # Run segmentation
+    labels = seg.fit_transform(data)
+
+    # Plot classification
+    mm.show_classification(labels, data)
+
+    # Plot classification distributions
+    mm.show_classification_distributions(labels, data)
+
+## NOTE Segmentation example using outliers
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: Using outlier removal

File: ssts/segmentation/example.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    # Run segmentation
+    labels = seg.fit_transform(data, outliers)
+
+    # Plot classification
+    mm.show_classification(labels, data)
+
+    # Plot classification distributions
+    mm.show_classification_distributions(labels, data)
+
+## NOTE Segmentation example with dimensionality reduction (PCA) across physical properties
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: Using dimensionality reduction across physical properties

File: ssts/segmentation/example.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    # Run segmentation
+    labels = seg.fit_transform(data)
+
+    # Plot classification
+    mm.show_classification(labels, data)
+
+    # Plot classification distributions
+    mm.show_classification_distributions(labels, data)
+
+## NOTE Segmentation example with dimensionality reduction (PCA) across neighboring pixels and physical properties
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: Using dimensionality reduction across neighboring pixels and physical properties

File: ssts/segmentation/segmentation_gmm.py [hunk 9]
```diff
@@ -112,20 +112,19 @@ def transform(self, data, outliers=None):
         else:
             data = data.reshape(n, c)
 
-        if outliers is not None:
-            outliers = outliers.flatten()
-
-        data = SegmenterGMM.remove_outliers(data, outliers)
```
Comment (dtorrejo) [inline ssts/segmentation/segmentation_gmm.py:118]: If we remove the outliers from the data, we won't be able to reconstruct the label matrix with dimensions (h, w). This didn't come up before because the dataset we were using didn't have outliers.

File: ssts/segmentation/segmentation_gmm.py [hunk 10]
```diff
@@ -112,20 +112,19 @@ def transform(self, data, outliers=None):
...
             data = self.sst.transform(data) if self.zscale else SegmenterGMM.normalize_by_max(data)
 
         if self.embedding_dim is not None:
             data = self.pca.transform(data)
 
         labels = self.gmm.predict(data)
+        labels = np.reshape(labels, (h, w))
+        labels += 1 # all labels move up one
+        if outliers is not None:
+            labels *= (1 - outliers) # outliers map to label 0
```
Comment (dtorrejo) [inline ssts/segmentation/segmentation_gmm.py]: I solve the issue addressed above by first classifying all the pixels in the original image, and then mapping the outlier mask to zeros

File: ssts/segmentation/example.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
+import numpy as np
+import segmentation_gmm as seg_gmm
+from methods import main_methods as mm # NOTE SSTS directory must be in PYTHONPATh
```
Comment (ponl) [inline ssts/segmentation/example.py]: Let go of the shift key too early: `PYTHONPATH`

File: ssts/segmentation/example.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    # Plot classification
+    mm.show_classification(labels, data)
+
+    # Plot classification distributions
+    mm.show_classification_distributions(labels, data)
+
+## NOTE Segmentation example with dimensionality reduction (PCA) across neighboring pixels and physical properties
+if which_example == 3:
+
+    data_dir = "/Users/diegotorrejon/Projects/Other/SSTS/data/2componentfilms/combined/PTB7PC71BM_CBonly_ascast_fresh_500 nm_combined.npy"
```
Comment (ponl) [inline ssts/segmentation/example.py]: You should abstract this out, especially since this is a public repo

Comment (dtorrejo) [reply ssts/segmentation/example.py]: Addressed

File: ssts/segmentation/example.py (hunk 11 above)
Comment (dtorrejo) [reply ssts/segmentation/example.py]: Lol. I fixed it anyway

**Comment by dtorrejo** [comment]
This PR should be good now.

File: methods/main_methods.py [hunk 13]
```diff
@@ -322,8 +322,12 @@ def show_classification_distributions(labels, data):
         labels (np.array): matrix of classification per pixel
         data (np.array): data
     """
+    unique_labels = list(np.unique(labels))
```
Comment (dtorrejo) [inline methods/main_methods.py]: Nicer way to figure out which labels to plot distributions for. The 0 label is avoided since it denotes outliers as well as the borders between grains (when using persistence watershed).

File: ssts/segmentation/example.py [hunk 14]
```diff
@@ -77,6 +78,25 @@ def main():
         # Plot classification distributions
         mm.show_classification_distributions(labels, data)
 
+    ## NOTE Segmentation example using persistence watershed on the height property of the material
+    elif example_number == "4":
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: Example using watershed segmentation on the height property of the material.

File: ssts/segmentation/example.py [hunk 15]
```diff
@@ -63,10 +63,22 @@ def main():
         # Plot classification distributions
         mm.show_classification_distributions(labels, data)
 
+        # Initialize Watershed segmentation
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: Applies watershed segmentation on output (labels) of GMM segmentation

File: ssts/segmentation/example.py [hunk 16]
```diff
@@ -43,10 +43,12 @@ def main():
         labels = seg.fit_transform(data, outliers)
 
         # Plot classification
-        mm.show_classification(labels, data)
+        no_outliers_data = np.copy(data)
+        no_outliers_data[outliers==1] = 0 # remove outliers from data
+        mm.show_classification(labels, no_outliers_data) # uses data without outliers
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: When removing outliers, we should do plots on data without outliers

File: ssts/segmentation/example.py [hunk 17]
```diff
@@ -110,6 +110,27 @@ def main():
         # Plot classification distributions
         mm.show_classification_distributions(labels, data)
 
+    ## NOTE Segmentation example using persistence watershed on the output of the sobel operator
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:146]: Applies watershed segmentation to the output of the sobel operator

File: ssts/segmentation/example.py [hunk 18]
```diff
@@ -54,28 +54,35 @@ def main():
...
 
         # Plot classification
-        #mm.show_classification(pre_labels, data)
+        mm.show_classification(pre_labels, data)
 
         # Plot classification distributions
-        #mm.show_classification_distributions(pre_labels, data)
+        mm.show_classification_distributions(pre_labels, data)
+
+        # Handling PCA components from data
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:77]: The PCA model is already trained if running the GMM segmentation class with a specified embedding_dim. So you can access its components in the following manner.

File: ssts/segmentation/example.py [hunk 19]
```diff
@@ -131,6 +138,27 @@ def main():
...
+    elif example_number == "6":
+
+        # Initialize GMM segmentation
+        pre_seg = seg_gmm.SegmenterGMM(n_components=2, embedding_dim=3)
+
+        # Get outliers
+        outliers = mm.extract_outliers(data)
+
+        # Remove height property
+        no_height_data = np.delete(data, 4, axis=2)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:180]: We can remove the height data from data using as shown above. Then we can use this to run the gmm segmentation class. This works well with the nano wires.

File: methods/main_methods.py [hunk 20]
```diff
@@ -319,28 +329,126 @@ def show_classification_distributions(labels, data):
...
+            ax_r.hist(data[:,:,i][labels == j], 50, alpha=0.8, density=True, color=CMAP(MAX_PIXEL - color_step))
 
-        pyplot.subplot(3,4,cnt)
-        pyplot.title(PROPS[i])
-        for j in unique_labels: # skip outliers
-            pyplot.hist(data[:,:,i][labels == j], 50, alpha=0.3, density=True)
+    pyplot.tight_layout()
+    pyplot.show()
+
+def show_grain_area_distribution(labels):
```
Comment (dtorrejo) [inline methods/main_methods.py:368]: Plots area (number of pixels) of segmented grains

File: methods/main_methods.py [hunk 21]
```diff
@@ -319,28 +329,126 @@ def show_classification_distributions(labels, data):
...
+    pyplot.plot(unique_labels, normal_grain_areas, 'b')
+    pyplot.xlabel('Label numbering')
+    pyplot.ylabel('Percentage in image')
+    pyplot.title('Grain Percentage in Image')
+    pyplot.grid()
+
+    pyplot.show()
 
+
+def show_distributions_together(labels, data):
```
Comment (dtorrejo) [inline methods/main_methods.py:405]: Plots all grains distributions under one plot

File: methods/main_methods.py [hunk 22]
```diff
@@ -319,28 +329,126 @@ def show_classification_distributions(labels, data):
...
+            color_step = int((index + 1) * MAX_PIXEL / len(grain_labels))
+            mask = np.ma.masked_where(labels!=j, MAX_PIXEL * np.ones(labels.shape) - color_step)
+            ax_l.imshow(mask, alpha=0.8, cmap=CMAP, aspect='auto', vmin=0, vmax=MAX_PIXEL)
+
+            ax_r.hist(data[:,:,i][labels == j], 50, alpha=0.8, density=True, color=CMAP(MAX_PIXEL - color_step))
 
     pyplot.tight_layout()
     pyplot.show()
 
+def show_distributions_separately(labels, data):
```
Comment (dtorrejo) [inline methods/main_methods.py:437]: Plots distributions of each grain per plot

File: methods/main_methods.py [hunk 23]
```diff
@@ -411,26 +411,32 @@ def show_distributions_together(labels, data):
...
             color_step = int((index + 1) * MAX_PIXEL / len(grain_labels))
+            colors_used.append(MAX_PIXEL - color_step)
             mask = np.ma.masked_where(labels!=j, MAX_PIXEL * np.ones(labels.shape) - color_step)
             ax_l.imshow(mask, alpha=0.8, cmap=CMAP, aspect='auto', vmin=0, vmax=MAX_PIXEL)
 
             ax_r.hist(data[:,:,i][labels == j], 50, alpha=0.8, density=True, color=CMAP(MAX_PIXEL - color_step))
 
     pyplot.tight_layout()
     pyplot.show()
+    return colors_used
```
Comment (dtorrejo) [inline methods/main_methods.py]: The colors are preserved to match the colors in show_distributions_separately method

File: ssts/segmentation/example.py [hunk 24]
```diff
@@ -4,21 +4,29 @@
...
-    p = argparse.ArgumentParser(description="""A script with examples on how to use segmentation code.""")
-    p.add_argument('data_path', help='Path to data.')
-    p.add_argument('example_number', help='Specifies which example to run.')
+from skimage import measure
+
+MAX_PIXEL = mm.MAX_PIXEL
+CMAP = mm.CMAP
+LABEL_THRESH = mm.LABEL_THRESH
+
+def main(data_path=None, example_number=None):
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:9]: @wesleyktatum you can now run this script using your spyder environment

File: ssts/segmentation/example.py [hunk 25]
```diff
@@ -76,21 +84,19 @@ def main():
         pca_components = pre_seg.pca.transform(data.reshape(n, c)) # NOTE PCA was trained while fitting pre_seg
         pca_components = pca_components.reshape(h, w, num_pca_components) # shape (512, 512, num_pca_components)
 
-        # Initialize Watershed segmentation # NOTE this module can be applied after any example
-        post_seg = seg_water.SegmenterWatershed(pers_thresh=0.3) # TODO why is it stochastic?
-
-        # Apply watershed segmentation on output of GMM segmentation
-        post_labels = post_seg.fit_transform(pre_labels)
+        # Create unique masks per grain
+        post_labels = measure.label(pre_labels, connectivity=2)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: This method is better at segmenting binary masks, as opposed to the watershed algorithm

File: ssts/segmentation/example.py [hunk 26]
```diff
@@ -107,17 +113,21 @@ def main():
...
         # Initialize GMM segmentation
         seg = seg_water.SegmenterWatershed()
 
         # Choose material property to segment
         prop_data = data[:,:,4] # NOTE height
 
         # Run segmentation
-        labels = seg.fit_transform(prop_data)
+        labels = seg.fit_transform(prop_data, outliers)
+        labels = measure.label(labels, connectivity=2) # creates unique labels
```
Comment (dtorrejo) [inline ssts/segmentation/example.py]: This ensures that there are unique labels per grain after watershed algorithm.

File: ssts/segmentation/segmentation_watershed.py [hunk 27]
```diff
@@ -86,7 +92,7 @@ def fit_transform(self, data, outliers=None):
 
     @staticmethod
     def normalize_data(data):
-        return (data - np.mean(data)) / np.std(data)
+        return np.abs(data) / np.max(np.abs(data))
```
Comment (dtorrejo) [inline ssts/segmentation/segmentation_watershed.py:101]: The normalized data ranges from 0 to 1. Thus a threshold of 0.5 is a "good" default parameter for the watershed algorithm.

File: ssts/segmentation/example.py [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+        seg = seg_water.SegmenterWatershed()
+
+        # Choose material property to segment
+        prop_data = data[:, :, 4]  # NOTE height
+
+        # Fit watershed algorithm
+        seg.fit(prop_data)
+
+        # Find optimal merging parameter
+        optimal_thresh = seg.find_optimal_thresh(prop_data, outliers, plot_flag=True)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:135]: First pass at getting optimal thresh for persistence watershed.

File: ssts/segmentation/example.py [hunk 29]
```diff
@@ -58,7 +58,7 @@ def main(data_path=None, example_number=None):
         mm.show_distributions_together(labels, no_outliers_data)  # all classes plotted together
         mm.show_distributions_separately(labels, no_outliers_data)  # each class plotted per plot
 
-    ## NOTE Segmentation example with dimensionality reduction (PCA) across physical properties
+    ## NOTE Segmentation example with dimensionality reduction (PCA) across physical properties AND grain segmentation
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:61]: Example 2 does PCA segmentation across properties (not using neighbor information) and does grain segmentation

File: ssts/segmentation/example.py [hunk 30]
```diff
@@ -96,17 +96,25 @@ def main(data_path=None, example_number=None):
     ## NOTE Segmentation example with dimensionality reduction (PCA) across neighboring pixels and physical properties
     elif example_number == "3":
 
+        # Get outliers
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:99]: Added outlier removal in example 3. This is optional

File: ssts/segmentation/example.py [hunk 31]
```diff
@@ -166,19 +174,37 @@ def main(data_path=None, example_number=None):
...
         # Plot classification
-        mm.show_classification(pre_labels, data)
+        mm.show_classification(labels, data)
 
         # Plot classification distributions
-        mm.show_classification_distributions(pre_labels, data)
+        mm.show_classification_distributions(labels, data)
+
+    ## NOTE Segmentation example with dimensionality reduction (PCA) across physical properties (skinny version of example 2)
+    elif example_number == "7":
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:192]: Skinny version of example 2 (i.e., without grain segmentation)

File: ssts/segmentation/segmentation_watershed.py [hunk 32]
```diff
@@ -97,7 +97,7 @@ def fit_transform(self, data, outliers=None, pers_thresh=DEF_THRESH):
         return self.transform(data, outliers, pers_thresh)
 
     @staticmethod
-    def normalize_data(data):  # Maps to (0,1). This allows negative values to be grains.
+    def normalize_data(data):  # Maps from (-a,b) to (0,1). This allows negative values to be grains.
```
Comment (dtorrejo) [inline ssts/segmentation/segmentation_watershed.py:100]: The height property has range from (-a, b). This normalization step maps the height to the interval (0,1). Also the background gets converted to a peak for the watershed algorithm to segment it.

**Comment by ponl** [comment]
lgtm
